### PR TITLE
[ir] Make BLSAnalyzer testable

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2177 

Hi @KLozes ,

Sorry this PR might cause more conflicts to your #2248 , but I believe this is a necessary step to allow us to better test the BLS code and make it more robust. Albeit the large change, it simply pulls out the `BLSAnalysis` class so that we can add test around it.

A few issues:

1. We don't see to handle this case correctly: when there are two GlobalPtrs, `BLSAnalyzer` can infer the BLS access pattern for one but not the other. Right now the scratch pads will be partially accessed, which seems wrong?
2. Make the interface of `BLSAnalysis` expose the result of whether BLS pads are initialized.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Review by xumingkuan — APPROVED** [review]
LGTM! Note that this test doesn't guarantee that accesses in the scratch pad's range are replaced with which to the BLS, but I think it's a good first step.

File: tests/cpp/analysis/bls_analyzer_test.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  void run(SNode &, bool) override {
+  }
+};
+
+constexpr int kBlockSize = 8;
+
+class BLSAnalyzerTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    const std::vector<Index> indices = {Index{0}, Index{1}};
```
[the remark this answers is not in the document]
Comment (k-ye) [reply tests/cpp/analysis/bls_analyzer_test.cpp:35]: Fixed

**Comment by k-ye** [comment]
> this test doesn't guarantee that accesses in the scratch pad's range are replaced with which to the BLS, but I think it's a good first step.

Right. I think this is what unit tests are supposed to do -- focus on the functionality of `BLSAnalyzer` itself (without caring about how pads are used later). It will be great if we have another test to cover what you've pointed out, though :-)

**Comment by k-ye** [comment]
@KLozes if you can make a separate PR that just changes how DiffRange multiplication is supported, I can defer this PR so as to reduce some merge conflict :-)

**Pull request merged**